### ISSUE TITLE
Please make master multithreaded

### ISSUE THREAD

**Issue by alexcrow** [moosefs/moosefs#11]
IMHO the biggest bottleneck in MooseFS is the single-threaded nature of the master server. We have about 320 roaming windows profiles on Samba+Moose and loading them is about 20-50% slower than on a single XFS hosted samba machine. During times when lots of users are logging in the master server CPU becomes almost saturated on the single master process.

If an organisation with 1000+ users was doing the same I should imagine it would be proportionally worse.

I appreciate this if probably not an easy task, so I won't hold my breath.

Regards

Alex

**Comment by acid-maker** [comment]
I agree that we need to speedup master, but I don't want to make it multithreaded. I'm thinking of spawning subprocesses to utilize more CPU cores. Such approach is much safer than using threads.

Also I see possibility to utilize master-followers for read-only operations (mainly lookup and getattr).

**Comment by davies** [comment]
Agreed with @acid-maker, multithread may not the good solution.

@alexcrow Could you show some stats from webui when the master is busy?

How many clients and chunkservers do you have? When there are hundreds of clients or chunkservers, the CPU cycles spent in network (and related bookkeeping) are not trivial, epoll should be used to serve more than 1000+ clients or chunkservers. I had done that before (also posted in the maillist), together with other optimization, it did decrease the CPU usage to 10-%.
